RFC-0001-torch-function-for-methods.md

```diff
@@ -0,0 +1,358 @@
+
+In addition, there has been interest in adding a "universal hook" that operated
+on both functions and methods, perhaps modifying the control flow before
+returning the result. Such a hook already exists today in the form of
+`__torch_function__`, however, it only operates on functions and not on
+methods, and support for subclassed `torch.Tensor` objects in this protocol is
+limited.
+
+### Proposal
+We propose the following signature change to `__torch_function__`, to make it
+match NumPy: [[4]]
+
+```python
+class SubTensor(torch.Tensor):
+    def __torch_function__(self, func, types, args, kwargs):
+        # Implementation here
+```
+
+The reason for adding `types` to the signature is necessitated by the need for
+`super()`. If we set a requirement for `super().__torch_function__` to work
+properly, we would need to provide an easy way for users to signal to
+`__torch_function__` that they are calling to the next-specific implementation.
+
+### Process followed during a function/method call
+
+The process followed during a function/method call would be equivalent to:
+
+1. The dispatcher is called to extract the `Tensor`-likes.
+2. All `Tensor`-likes are checked for `__torch_function__`. If none exist, the
+   internal implementation is called, and the final result is returned.
+3. A collection of types that implement `__torch_function__` is created, with
+   no guaranteed order other than that subclasses come before superclasses.
+4. For one instance of each type in `types`, `__torch_function__` is called.
+   The first such function or method to return something other than
+   `NotImplemented` will be the final result. All exceptions will be propagated
+   upward.
+5. If all `__torch_function__` implementations return `NotImplemented`, a
+   `TypeError` is raised with an appropriate error message.
+
+In practice, for most PyTorch functions, the list of tensor-likes is already
+available and the dispatcher doesn't need to be called. Additionally, while
+equivalent to the code above, if the `Tensor`-likes are all `Tensor` or don't have
+an `__torch_function__` implementation, the internal implementation is called
+immediately. This is done as a performance optimisation to avoid overhead for
+concrete `Tensor` objects.
+
+It will be the job of the dispatcher to extract `Tensor`-like objects from the
+argument list, however, arguments of type `Optional[Tensor]` will be considered
+`Tensor`-like. If one gets a compound or dependent type such as `List[Tensor]`
+or `Tuple[Tensor, ...]` or `Tuple[Tensor, int]`, the dispatcher will have the job
+of extracting an iterable of objects that *could* be `Tensor`-like.
+
+### Generic implementation of `__torch_function__`
+`torch.Tensor` will gain a generic `__torch_function__` of the following form:
+
+```python
+class Tensor:
+    def __torch_function__(self, func, types, args, kwargs):
+        if not all(issubclass(type(self), t) for t in types):
+            return NotImplemented
```

Comment:
If we eliminate `types` as an explicitly passed in argument, the modified test would simply extra tensor-likes from args and then check that the types are consistent. We do this already to determine what `__torch_function__` to call, so I think for the internal implementation this would not be a big deal to do. I don't *think* that users need to do this, but I'm willing to be convinced otherwise.

Reply:
The main reason for having the `types` in here was threefold:

* Do not mix unrelated class trees. Removing this would.
* Provide an easy way for a `__torch_function__` implementation to check if it can support the supplied types.
* Signal that dispatch has to go through a different path (now defunct, so I agree with you).

Expanding on each of these a little,

### Do not mix unrelated class trees
Suppose there is `Tensor`, `SubTensorA`, and `SubSubTensorA` that form an inheritance chain, and `Tensor`, `SubTensorB` and `SubSubTensorB` that form another inheritance chain. If we remove this class, `SubTensorA + SubSubTensorA` would return `SubSubTensorA` (good, subclasses come before superclasses), but `SubTensorB + SubTensorA` would return `SubTensorB` (oookay...) and `SubTensorA + SubTensorB` would return `SubTensorA` (a-ha! inconsistency found!)

With this part of the proposal in place, both of the above would raise `TypeError`.

This applies not just to `Tensor` subclasses but also all kinds of mixing. See below.

### Check for support
Right now, the `Tensor`-likes are buried inside the `args` and `kwargs`. Now, envision a future where we have implementations of `__torch_function__` on a lot of classes, possibly being mixed.

We faced exactly this dilemma in `__array_function__`, and `types` was the solution.

Assume `DuckTensorA` and `DuckTensorB` are two such classes. Now envision a fairly complex function like `torch.cat`, or the equivalent of [`numpy.block`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.block.html#numpy.block).

How are `DuckTensorA` and `DuckTensorB` to know that they don't support each other, without going recursively into each argument and extracting the type (which is fatal for performance, btw)? Why, they just inspect `types`!

Reply:
> Do not mix unrelated class trees

I agree that we should not mix unrelated class trees. I wonder if the "reference" implementation on Tensor (that rewraps subclasses) can directly determine if class tree mixing would occur by looking at the arguments. It already needs to extract types from the tensor like arguments because we need to determine which `__torch_function__` to invoke for multiple dispatch; so it can also easily check if all the subclasses are related or not. Is there a use case for user-defined `__torch_function__` to also do this calculation?

> Check for support

Yeah, no counterargument here. :) Alright, types can stay :> (however, I'm still not convinced you have to strip out types below you in the subclass hierarchies on super calls!)